internal/backend/runtime/omni/controllers/omni/node_unique_token_status.go

```diff
@@ -0,0 +1,211 @@
+			},
+			TransformFunc: handler.reconcileRunning,
+		},
+		qtransform.WithExtraMappedInput(qtransform.MapperSameID[*siderolinkres.NodeUniqueToken, *machineStatusLabels]()),
+		qtransform.WithExtraMappedInput(qtransform.MapperSameID[*omni.ClusterMachine, *machineStatusLabels]()),
+		qtransform.WithExtraMappedInput(qtransform.MapperSameID[*omni.MachineStatusSnapshot, *machineStatusLabels]()),
+		qtransform.WithExtraMappedInput(qtransform.MapperSameID[*omni.Machine, *machineStatusLabels]()),
+		qtransform.WithExtraMappedInput(mappers.MapClusterResourceToLabeledResources[*omni.TalosConfig, *machineStatusLabels]()),
+		qtransform.WithConcurrency(32),
+	)
+}
+
+type nodeUniqueTokenStatusHandler struct{}
+
+func (handler *nodeUniqueTokenStatusHandler) reconcileRunning(
+	ctx context.Context,
+	r controller.Reader,
+	logger *zap.Logger,
+	machineStatus *machineStatusLabels,
+	nodeUniqueTokenStatus *siderolinkres.NodeUniqueTokenStatus,
+) error {
+	talosVersion, ok := machineStatus.Metadata().Labels().Get(omni.MachineStatusLabelTalosVersion)
+
+	if !ok {
+		logger.Debug("the token is not generated: the machine version is unknown")
+
+		return nil
+	}
+
+	if !siderolink.SupportsSecureJoinTokens(talosVersion) {
+		logger.Debug("the token is not generated: not supported by the machine Talos version",
+			zap.String("version", talosVersion),
+		)
+
+		nodeUniqueTokenStatus.TypedSpec().Value.State = specs.NodeUniqueTokenStatusSpec_UNSUPPORTED
+
+		return nil
+	}
+
+	token, err := safe.ReaderGetByID[*siderolinkres.NodeUniqueToken](ctx, r, machineStatus.Metadata().ID())
+	if err != nil && !state.IsNotFoundError(err) {
+		return err
+	}
+
+	_, installed := machineStatus.Metadata().Labels().Get(omni.MachineStatusLabelInstalled)
+
+	if token != nil && token.TypedSpec().Value.Token != "" {
+		if installed {
+			nodeUniqueTokenStatus.TypedSpec().Value.State = specs.NodeUniqueTokenStatusSpec_PERSISTENT
+		} else {
+			nodeUniqueTokenStatus.TypedSpec().Value.State = specs.NodeUniqueTokenStatusSpec_EPHEMERAL
+		}
+
+		// no need to do anything, the token is already generated
+		return nil
+	}
+
+	nodeUniqueTokenStatus.TypedSpec().Value.State = specs.NodeUniqueTokenStatusSpec_NONE
+
+	if _, connected := machineStatus.Metadata().Labels().Get(omni.MachineStatusLabelConnected); !connected {
+		logger.Debug("the token is not generated: the machine is not connected")
+
+		return nil
+	}
+
+	c, err := handler.getClient(ctx, r, machineStatus.Metadata().ID())
+	if err != nil {
+		return err
+	}
+
+	defer c.Close() //nolint:errcheck
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+
+	logger.Info("proactively generating node unique token")
+
+	return handler.generateUniqueNodeToken(ctx, c, logger)
+}
+
+// getMachineFingerprint gets all network devices and calculates the checksum out of their mac addresses.
+func (handler *nodeUniqueTokenStatusHandler) getMachineFingerprint(ctx context.Context, c *client.Client) (string, error) {
```

Comment:
should we reuse this from `pendingMachineStatusHandler`? move to a common place?